### ISSUE TITLE
NamedExec batch INSERT does not work with ON CONFLICT

### ISSUE THREAD

**Issue by bretep** [jmoiron/sqlx#505]
Recently commited feature #285 does not support ON CONFLICT.

example:
```
sots := []*Sot{&sot1, &sot2}
db.NamedExec("INSERT INTO sot (sot_identifier, name)
			VALUES (:sot_identifier, :name)
			ON CONFLICT (sot_identifier) DO
				UPDATE SET (name) = (
				EXCLUDED.name)", sots)
```

Here is the query generated by `NamedExec`:
```
INSERT INTO sot (sot_identifier, name)
			VALUES ($1, $2)
			ON CONFLICT (sot_identifier) DO
				UPDATE SET (name) = (
				EXCLUDED.name),(
				EXCLUDED.name)
```

**Comment by bretep** [comment]
@hmgle is this something you could fix easily?

**Comment by bretep** [comment]
Looks like `sqlx.Named` also has the same issue.

**Comment by hmgle** [comment]
This bug may be caused by the `fixBound` function. I will check it later.

**Comment by hmgle** [comment]
This line [EndBracketsReg = regexp.MustCompile(`\([^()]*\)\s*$`)](https://github.com/jmoiron/sqlx/blob/cdf62fdf55f66fb2484fe214f89c059f0bd3f567/named.go#L214) assumes that the `value_list` part is at the end of `bound`. We cat fix the regexp like this:

```go
var valueBracketReg = regexp.MustCompile(`\([^(]*\?+[^)]*\)`)

func fixBound(bound string, loop int) string {
	loc := valueBracketReg.FindStringIndex(bound)
	if len(loc) != 2 {
		return bound
	}

	var buffer bytes.Buffer

	buffer.WriteString(bound[0:loc[1]])
	for i := 0; i < loop-1; i++ {
		buffer.WriteString(",")
		buffer.WriteString(bound[loc[0]:loc[1]])
	}
	buffer.WriteString(bound[loc[1]:])
	return buffer.String()
}
```

The patch:

```diff
-var (
-	EndBracketsReg = regexp.MustCompile(`\([^()]*\)\s*$`)
-)
+var valueBracketReg = regexp.MustCompile(`\([^(]*\?+[^)]*\)`)
 
 func fixBound(bound string, loop int) string {
-	endBrackets := EndBracketsReg.FindString(bound)
-	if endBrackets == "" {
+	loc := valueBracketReg.FindStringIndex(bound)
+	if len(loc) != 2 {
 		return bound
 	}
+
 	var buffer bytes.Buffer
-	buffer.WriteString(bound)
+
+	buffer.WriteString(bound[0:loc[1]])
 	for i := 0; i < loop-1; i++ {
 		buffer.WriteString(",")
-		buffer.WriteString(endBrackets)
+		buffer.WriteString(bound[loc[0]:loc[1]])
 	}
+	buffer.WriteString(bound[loc[1]:])
 	return buffer.String()
 }
```

**Comment by smartfly** [comment]
good

**Comment by asadhu** [comment]
@bretep we are facing a similar issue using ON CONFLICT, which release would have this fix?

**Comment by mai00cmi** [comment]
Looks like b63be726 reverted this fix and now ON CONFLICT is broken again.

**Comment by aarengee** [comment]
How to do batch upserts workaround with 1.3.0+ unless fixed later.
https://github.com/jmoiron/sqlx/pull/667#issuecomment-789656681

**Comment by os** [comment]
I believe this is fixed in 1.3.3.
